### PR TITLE
Add large cactus seedling

### LINKED ISSUE #2255: Cactus seedling
In the current version of Minetest, the Cactus naturally grows in the shape of
cactus
Where as grown cactus looks like this:
screenshot_2018-10-28-16-46-15-1
So there is no way to re-create the original look
So I think there should be a cactus seedling!
big cactus
screenshot_2018-10-28-16-36-49-1
### /\ SUBJECT TO CHANGE /\
raw file
t026

### PR DESCRIPTION
**Pull request by paramat**
screenshot_20181129_080229

screenshot_20190101_023409

^ Tested the updated chat message sent to player if a placed seedling will intersect protection, now the sapling/seedling description is used.

Closes #2255

### CODE REVIEW

**Comment by Extex101** [comment]
Thanks paramat 
Will this be added to the Minetest wiki?

**Comment by Extex101** [comment]
:+1:

File: mods/default/crafting.lua [hunk 1]
```diff
@@ -779,6 +779,15 @@ minetest.register_craft({
 	}
 })
 
+minetest.register_craft({
+	output = "default:large_cactus_seedling",
+	recipe = {
+		{"", "default:cactus", ""},
+		{"default:cactus", "default:cactus", "default:cactus"},
+		{"", "default:cactus", ""},
```
Comment (Ezhh) [inline mods/default/crafting.lua]: Maybe a bit too expensive?

**Comment by Ezhh** [comment]
Final timer duration needs to average around the time taken to grow x number of new cactus nodes the standard way, where x is total nodes per large cactus - however many the recipe costs to make a sapling.

File: mods/default/crafting.lua (hunk 1 above)
Comment (paramat) [reply mods/default/crafting.lua]: Originally it was 2 cactus nodes, but that results in a very large number of seedlings per cactus. This way you get 2 from a large cactus which seems right, it also visually looks like the large cactus.

**Comment by paramat** [comment]
>  however many the recipe costs to make a sapling.

Ok agreed.

**Comment by paramat** [comment]
Most comments attended too, sorry i squashed the commits, forgot.
The timer duration needs to be chosen through experiment, as guessing ABM behaviour has proven to be wildly inaccurate in the past.

**Comment by paramat** [comment]
Placement issues ...
Because the large cactus schematic is unequal in x and z dimensions the seedling is not where the trunk is. So far i have tried:
* Seedling at the centre of a 5x5 square where the large cactus is placed along the South or West edges.
* Seedling under one end of the large cactus.

However this can be solved if we make the cactus 5x5 but mostly air, with the cactus as the central slice.

**Comment by paramat** [comment]
Done. The non-'square' schematic was also causing problems for the protection intersection check.
Now i just need to tune the timer duration by experiment. Everything else is non-WIP.

**Comment by Ezhh** [comment]
Placement should really be calculated after rotation is decided so that this isn't needed. Are the place_center flags not working correctly?

This came up with goldleaf trees in HW as well, so there should be a better solution.

**Comment by paramat** [comment]
Ezhh, indeed it's failing of the engine, but altering behaviour may break mods. I might look into this after 5.0.0 release. However, with a 5x7x5 cactus we could have more branches and not be a flat shape.
The 'on place' protection check cannot know which rotation will be used, so will always need to be a 5x7x5 volume.

Extex101 that uses replacements to remove the seedling, which is a bit hacky, i'd rather 'remove', however now this is not needed as the trunk appears where the seedling is.

**Comment by paramat** [comment]
PR complete. Growth time tuned to be no faster, by comparing crafting 5 nodes into a seedling against farming 5 nodes.
```
		-- Normal cactus farming adds 1 cactus node by ABM,
		-- interval 12s, chance 83.
		-- Consider starting with 5 cactus nodes. We make sure that growing a
		-- large cactus is not a faster way to produce new cactus nodes.
		-- Confirmed by experiment, when farming 5 cacti, on average 1 new
		-- cactus node is added on average every
		-- 83 / 5 = 16.6 intervals = 16.6 * 12 = 199.2s.
		-- Large cactus contains on average 14 cactus nodes.
		-- 14 * 199.2 = 2788.8s.
```
Growth time is 46 mins, almost twice the maximum tree growth time.

**Comment by Extex101** [comment]
Will this be merged?

**Comment by paramat** [comment]
Extex101 please don't add unnecessary comments, some PRs take months to be reviewed.

**Comment by paramat** [comment]
> Consider all nodes that possibly are in the way of other nodes as half (probably even less) value (when one cactus can't grow into a space because there's already a cactus there).

That's the player's fault if they plant them too close to each other, they deserve the result. With a shorter grow time the players that don't plant them too close too each other (most players) will then have a faster method than standard farming.

**Comment by dopik** [comment]
Then to my second point:
Using your calculations but considering that the player immidiatly replants the cacti he gets, the average grow time per seedling gets reduced to about 18mins.
See: 
>EDIT: if you immidiatly replant the cacti you get (in your calculation) you get an average growtime for the seedling of 1088.4 s = 18 mins. Calculated by adding 83/5 + 83/6 + etc. + 83/13= 90.7 . Multiply that by 12s

To make the calculation clearer: 83/5 is the average cycle time for one cactus to grow( as used in your calculation). After that first cactus we now have 6 cacti, so for the second cactus to grow it now only takes 83/6 to grow, and so on... resulting in an average of 90.7 cycles (as in abm cycles) to get 14 cacti.

**Comment by paramat** [comment]
Ok, but that assumes:
> if you immidiatly replant the cacti you get

This won't happen in all situations, and we have to go by the slower classic farming method.

Just noticed our avatars are similar.

**Comment by dopik** [comment]
Ok, my last argument:
 If we were to use a grow time where the seedling and normal cacti produce the same amount, then the seedling would be worse in value, because it needs more space. To account for this, tuning down the grow time a bit from the equilibrium point between cactus and seedling (=same production per time) is necessary, IMO

**Comment by Ezhh** [comment]
The seedling is meant to be the same **or worse** in value. Certainly not better. The existing method of farming cacti is **farming** cacti , where as this is restoring the option to have them grow as they do in the wild. It makes complete sense that space-efficient method of growth used for farming purposes would produce more.

**Comment by paramat** [comment]
I agree with Ezhh.
Space isn't taken into account in my calculations as it's usually a non-issue in a desert. Those who aren't concerned by space would then have a faster method than classic farming.

**Comment by paramat** [comment]
If it is preferred that i add a 3rd cactus branch, to make the 5x7x5 schematic more used, let me know.

**Comment by paramat** [comment]
@Ezhh @SmallJoker no huge rush but i'd like to get this into 5.0.0 release.

File: mods/default/trees.lua [hunk 2]
```diff
@@ -550,7 +560,8 @@ function default.sapling_on_place(itemstack, placer, pointed_thing,
 			interval) then
 		minetest.record_protection_violation(pos, player_name)
 		-- Print extra information to explain
-		minetest.chat_send_player(player_name, "Tree will intersect protection")
+		minetest.chat_send_player(player_name, "Will intersect protection " ..
```
Comment (SmallJoker) [inline mods/default/trees.lua]: `itemstack:get_definition().description` would be helpful here.

**Comment by paramat** [comment]
As requested will add randomness to timer.

**Comment by paramat** [comment]
@SmallJoker done.

File: mods/default/trees.lua (hunk 2 above)
Comment (paramat) [reply mods/default/trees.lua]: Done.

**Comment by paramat** [comment]
As requested by SmallJoker on IRC i have added a random range to the timer, averaging to the previous time.

**Comment by paramat** [comment]
Tested. See new screenshot in first post for chat message update.

**Pull request merged**